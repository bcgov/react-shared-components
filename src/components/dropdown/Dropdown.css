.bcgov-form {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.bcgov-dropdown-label {
  margin-bottom: 10px;
}

.bcgov-dropdown {
  color: #313132;
  background: white;
  box-shadow: none;
  border: 2px solid #606060;
  width: 100%;
  padding: 8px 45px 8px 15px;
  appearance: none;
}

.fa-chevron-down {
  pointer-events: none;
  position: absolute;
  top: 1em;
  right: 1em;
}

.bcgov-dropdown-wrapper {
  position: relative;
  display: inline;
}

:focus {
  outline: 4px solid #3b99fc;
  outline-offset: 1px;
}
