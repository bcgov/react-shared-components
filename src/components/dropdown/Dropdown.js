import React from "react";
import PropTypes from "prop-types";
import { FaChevronDown } from "react-icons/fa";

import "./Dropdown.css";

export const Dropdown = ({ items, label, onSelect }) => {
  return (
    <div className="bcgov-form">
      {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
      <label className="bcgov-dropdown-label" htmlFor="dropdown">
        {label}
      </label>
      <div className="bcgov-dropdown-wrapper">
        <FaChevronDown className="fa-chevron-down" />
        <select
          className="bcgov-dropdown"
          name="options"
          id="dropdown"
          data-testid="dropdown"
          onChange={(e) => onSelect(e.target.value)}
        >
          {items.map((val) => (
            <option key={val} value={val}>
              {val}
            </option>
          ))}
        </select>
      </div>
    </div>
  );
};

Dropdown.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  label: PropTypes.string.isRequired,
  onSelect: PropTypes.func.isRequired,
};
