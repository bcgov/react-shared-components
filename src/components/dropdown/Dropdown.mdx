import { Meta, Story } from "@storybook/addon-docs";
import { Dropdown } from "./Dropdown";

<Meta title="Dropdown" component={Dropdown} />

# Dropdown

Dropdowns allow users to select one option from a list.

## When to Use

When users have to select one option from a list of seven (7) to fifteen (15) different options.

## When not to Use

- When users have to select one option from a list of less than seven (7) items. Consider using a Radio Button instead
- When users have to select multiple options from a list. Use Checkboxes instead
- When users have to select one option from a very long list such as their country of birth. Consider using a Text Input that suggests input options instead

## API

| Prop     | Type     | Required | Description                                                      | Default |
| -------- | -------- | -------- | ---------------------------------------------------------------- | ------- |
| label    | string   | false    | value of the heading label for the dropdown                      | ""      |
| items    | array    | true     | an array of strings representing the items in the dropdown       |         |
| onSelect | function | true     | function describing action to take on selection of dropdown item |         |

<br />

## With Pre-Selected Option

Basic dropdown with a pre-selected option.

<Story
  name="With Pre-selected Option"
  id="dropdown--pre-selected-option"
></Story>

```bash
const items = ["Option 1", "Option 2", "Option 3", "Option 4"];

<Dropdown label="My dropdown" items={items} onSelect={onSelect} />
```

## With No Pre-Selected Option

Basic dropdown with no pre-selected option.

<Story
  name="With No Pre-selected Option"
  id="dropdown--no-pre-selected-option"
></Story>

```bash
const items = ["Select", "Option 1", "Option 2", "Option 3", "Option 4"];

<Dropdown label="My dropdown" items={items} onSelect={onSelect} />
```

## With No Label

Basic dropdown with no label.

<Story name="With No Label" id="dropdown--no-label"></Story>

```bash
const items = ["Select", "Option 1", "Option 2", "Option 3", "Option 4"];

<Dropdown items={items} onSelect={onSelect} />
```
