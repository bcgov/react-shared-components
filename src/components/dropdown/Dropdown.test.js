import React from "react";
import { render, fireEvent, getByTestId } from "@testing-library/react";
import { Dropdown } from "./Dropdown";
import testBasicSnapshot from "../../TestHelper";

describe("Dropdown Component", () => {
  const label = "My dropdown";
  const items = ["Option 1", "Option 2"];
  const onSelect = jest.fn();

  test("matches the snapshot", () => {
    testBasicSnapshot(
      <Dropdown onSelect={onSelect} label={label} items={items} />
    );
  });

  test("on change updates the selected value", () => {
    const { container } = render(
      <Dropdown onSelect={onSelect} label={label} items={items} />
    );

    fireEvent.change(getByTestId(container, "dropdown"), {
      target: { value: "Option 2" },
    });

    expect(onSelect).toHaveBeenCalledWith("Option 2");
  });
});
