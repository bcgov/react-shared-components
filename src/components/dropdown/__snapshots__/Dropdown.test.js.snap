// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component matches the snapshot 1`] = `
<div
  className="bcgov-form"
>
  <label
    className="bcgov-dropdown-label"
    htmlFor="dropdown"
  >
    My dropdown
  </label>
  <div
    className="bcgov-dropdown-wrapper"
  >
    <svg
      className="fa-chevron-down"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
      />
    </svg>
    <select
      className="bcgov-dropdown"
      data-testid="dropdown"
      id="dropdown"
      name="options"
      onChange={[Function]}
    >
      <option
        value="Option 1"
      >
        Option 1
      </option>
      <option
        value="Option 2"
      >
        Option 2
      </option>
    </select>
  </div>
</div>
`;
